app/vminsert/netstorage/netstorage.go

```diff
@@ -671,20 +671,14 @@ func initStorageNodes(addrs []string, hashSeed uint64) *storageNodesBucket {
 
 				if sn.isReady() {
 				again:
-					oldSnb := getStorageNodesBucket()
-					snbNew := storageNodesBucket{
-						ms:     oldSnb.ms,
-						wg:     oldSnb.wg,
-						stopCh: oldSnb.stopCh,
-						sns:    oldSnb.sns,
-					}
-
-					newNodeIDs := append(oldSnb.nodesHash.nodeHashes, sn.getID())
-					snbNew.nodesHash = newConsistentHash(newNodeIDs, hashSeed)
-
-					if !storageNodes.CompareAndSwap(oldSnb, &snbNew) {
+					currentSnb := getStorageNodesBucket()
+					newSnb := initStorageNodes(addrs, hashSeed)
```

Comment:
If the `sn` changes to ready, `initStorageNodes` will be executed. I noticed that `initStorageNodes` contains some code for metrics registration（`ms.NewGauge(fmt.Sprintf(`vm_rpc_rows_pending{name="vminsert", addr=%q}`, addr)`）. Will multiple registrations cause any issues (like panic)? Perhaps adding a new method `updateStorageNodes` would be better.

Reply:
Hello @shichanglin5 

> I noticed that initStorageNodes contains some code for metrics registration（'ms.NewGauge(fmt.Sprintf(vm_rpc_rows_pending{name="vminsert", addr=%q}, addr)'）. Will multiple registrations cause any issues (like panic)? 

No, this will not cause panic as long as metrics belong to different `MetricSet`s.
 
> Perhaps adding a new method `updateStorageNodes` would be better. 

It would be more efficient, but it makes code much more complicated. It will require to introduce synchronization for `snb.sns` and `snb.nodesHash`, which means that there will be performance penalty at the hot path. Also it will make the code more complicated as all functions that use `snb` will require to acquire a lock (or an atomic copy of `sns` and/or `nodesHash`).
Considering that the case when storage nodes are not available on vminsert start is quite rare (it is always recommended to start storage nodes first) it seems like it efforts of doming more efficient update just don't make it.